File: 0011-container-with-most-water/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    let area = 0;
    let i = 0, j = height.length - 1;
    
    while(i < j) {
        const h = Math.min(height[i], height[j]);
        area = Math.max(area, h * (j - i));
        // find the next taller line
        if (height[i] < height[j]) i++;
	    else j--;
    }
    
    return area;
};
// tc -> O(n)
// sc -> O(1)

File: 0011-container-with-most-water/0011-container-with-most-water.go
func maxArea(height []int) int {
    n := len(height)
    l := 0
    r := n - 1
    res := 0
    tArea := 0
    
    for l < r {
        // calculate the current area 
        mHeight := min(height[l], height[r])
        tArea = mHeight * (r - l)
        
        if tArea > res {
            res = tArea
        }
        
        // to maximize area we need to update l and r respectively
        if(height[l] > height[r]) {
            r--
        } else {
            l++
        }
    }
    
    return res
}